core/src/main/scala/kafka/log/OffsetIndex.scala

```diff
@@ -234,8 +244,49 @@ class LazyOffsetIndex(@volatile private var _file: File, baseOffset: Long, maxIn
   }
 
   def get: OffsetIndex = {
+    if (isClosed)
+      throw new IllegalStateException(s"Attempt to access the closed OffsetIndex (file=${_file}, baseOffset=${baseOffset}, " +
+                                      s"maxIndexSize=${maxIndexSize}, writable=${writable}.")
     if (offsetIndex.isEmpty)
       offsetIndex = Some(new OffsetIndex(_file, baseOffset, maxIndexSize, writable))
     offsetIndex.get
   }
+
+  /**
+   * Close this index file.
+   * Note: This will be a no-op if the index has already been closed.
+   */
+  def close(): Unit = {
+    if (!isClosed) {
+      offsetIndex.foreach(_.close())
+      isClosed = true
+    }
+  }
+
+  /**
+   * Delete the index file that backs this offset if exists.
+   * This method ensures that if the index file has already been closed, it will not be recreated as a side effect.
+   *
+   * @throws IOException if deletion fails due to an I/O error
+   * @return `true` if the file was deleted by this method; `false` if the file could not be deleted because it did
+   *         not exist
+   */
+  def deleteIfExists(): Boolean = {
+    if (isClosed)
+      Files.deleteIfExists(file.toPath)
+    else
+      get.deleteIfExists()
+  }
+
+  /**
+   * Rename the file that backs this offset index if the index has ever been initialized or file already exists.
+   *
+   * @throws IOException if rename fails for defined index or existing file.
+   */
+  def renameTo(f: File) {
+    try {
+      if (offsetIndex.isDefined || file.exists)
```

Comment:
why do we need this check? Should we just use the renameTo in AbstractIndex to keep the same behavior?

Reply:
Using `renameTo` in `AbstractIndex` requires creation of an `OffsetIndex` (similarly a `TimeIndex`). Hence, in case the `OffsetIndex` or `TimeIndex` has never been constructed (i.e. it was being used as lazy offset/time index), this call redundantly creates an `OffsetIndex` or `TimeIndex`, which incurs an expensive memory mapping operation.

Using this lazy `renameTo` ensures that this expensive memory mapping operation will be avoided whenever possible.